### ISSUE TITLE
Suggestion: Remove package-lint--check-package-version-present

### ISSUE THREAD

**Issue by meedstrom** [purcell/package-lint#275]
I propose that `package-lint--check-package-version-present` should no longer fire.

Only the GNU/NonGNU Elpa repositories require a Package-Version header, and in Melpa there is a case to be made that the absence is not a bug but a feature, reducing the incidence of mistakes.  If emacs-devel wants to see the header, they can probably detect the absence themselves.

It's not a big deal, of course, since this only counts as a warning and not an error. However the sheer number of warnings I have had to ask for exceptions from is getting pretty high, and at some point the MELPA maintainers may go by feel of the number of warnings excepted-so-far.  Related: https://github.com/melpa/melpa/pull/9133#issuecomment-2336801635

**Comment by purcell** [comment]
Responded here:

[At least for single file packages, you can't install them without a `Package-Version` header.](https://github.com/melpa/melpa/pull/9133#issuecomment-2352503339)

**Comment by meedstrom** [comment]
Alright, I would regard that a bug in package.el, because it breaks what needn't be broken. I'll probably write to bug-gnu-emacs later.

I have a reservation about adding a `;; Package-Version: 0`, which is that I do not know if new package managers like Elpaca might not then favour that over the git/hg tag, once it is present.  Omitting it sends a clear message.  

I suppose I could go look up how these package managers actually work, but a new one could always appear, and it is strange to be required to add a non-semantic line in the first place.

**Comment by meedstrom** [comment]
I surmise that multi-file packages work because MELPA builds generate a `-pkg.el` file? If so, maybe it can also insert a `;; Package-Version: 0` in the single-file packages :) 

And then the way is clear to implement this suggestion. (Still need to write to bug-gnu-emacs, of course, anyway)

**Comment by purcell** [comment]
> Alright, I would regard that a bug in package.el, because it breaks what needn't be broken. I'll probably write to bug-gnu-emacs later.

I don't think it's a bug, thought. It's reasonable (and common in other ecosystems, e.g. debian packages) to install a package from a file that has been obtained separately from some source, and for that file to be self-describing, including containing a version. For single-file packages, `Package-Version` is where that version is placed, and that doesn't seem like a bad thing.

> I surmise that multi-file packages work because MELPA builds generate a `-pkg.el` file? If so, maybe it can also insert a `;; Package-Version: 0` in the single-file packages :)

MELPA already inserts the (correct) version into the `Package-Version` header of single-file packages. Again, because they were ultimately installed via `package-install-file`, which requires that metadata.

**Comment by meedstrom** [comment]
Yes, I suppose it is only a bug in some sense, perhaps I should not throw around the word so lightly.

> MELPA already inserts the (correct) version 

That's great, and even better than inserting a Version 0!

I think if the repo has tags, it is self-describing, no file needs to be, but yea, it makes sense that when pulling files out of the repo to make a release, a string-substitution is performed somewhere. 

Since MELPA is already doing this, I guess any problems would mainly be encountered when someone attempts `package-install-file` manually on a manually cloned repo. I take it that you have opted against just... leaving it to `package-install-file` to print the error in that situation?

I'd be a bit more iconoclastic, but that has pros and cons. You can close the issue. :)

This has actually enlightened me on what Straight/Elpaca do not do, so thanks... Currently Elpaca is in a complicated situation because it wants to warn about mismatched dependencies, but also default to doing a shallow clone, in which the tag information is gone, and since the cloned repo is something the user is supposed to be able to develop inside, it can do no string substitutions.  My workaround now as a packager is to [always keep a tag on the distribution branch](https://github.com/progfolio/elpaca/issues/353#issuecomment-2352935785).

I wonder if Git can include the tag history even without including the commits... maybe a "LAST_TAG" file can just be inserted into `.git/`...

**Comment by meedstrom** [comment]
> FWIW, MELPA also uses shallow clones, but for MELPA Stable we grab all the tags, determine the highest-numbered version tag, and then check that out.

This requires cloning the full repo first though, right? Pardon, I haven't tried it.

**Comment by meedstrom** [comment]
Ah, I see, there is a `git ls-remote` thing!

**Comment by purcell** [comment]
Oh nice. I actually think we clone the full repo in that case, but we can perhaps streamline it.

**Comment by meedstrom** [comment]
In case you want to see an example, I just did a PR to Elpaca.

https://github.com/progfolio/elpaca/pull/359/commits/07b8c111b1be3abfaf3aee6ad9cc60893be4be9f

Rewriting in self-contained vanilla Elisp, it looks like this:

```elisp
(with-temp-buffer
  (if (zerop (call-process "git" nil t nil
                           "-c" "versionsort.suffix=-" ;; https://git-scm.com/docs/git-config#Documentation/git-config.txt-versionsortsuffix
                           "ls-remote" "--refs" "--tags"
                           "--exit-code" ;; https://git-scm.com/docs/git-ls-remote.html#Documentation/git-ls-remote.txt---exit-code
                           "--sort=-version:refname" ;; https://git-scm.com/docs/git-tag#Documentation/git-tag.txt---sortltkeygt
                           "origin"))
      (mapcar (lambda (line)
                (thread-first line
                              (split-string "\t" t)
                              (cadr)
                              (split-string "/" t)
                              (caddr)))
              (split-string (buffer-string) "\n" t))
    (message "Git failed with output:\n %s" (buffer-string))))
```

and if you eval that in a directory that is a git repo, you should get the version-sorted list of tags, even if that repo is shallow.

**Comment by alphapapa** [comment]
FWIW, note that the Elisp manual says:
```
   The version number comes from the ‘Package-Version’ header, if it
exists, or from the ‘Version’ header otherwise.  One or the other _must_
be present.
```
I was initially confused since I've always used `Version:` rather than `Package-Version:`.

And FWIW, I would not like this check to be disabled by default, because then I'd have to configure all of my packages to specifically enable it (and invent a way for [makem.sh](https://github.com/alphapapa/makem.sh) to pass such an option through to `package-lint`).

@meedstrom When considering changes to widely used tools, it's very important to consider what effects the changes could have on downstream users and tools and whether such changes are really necessary.  Sometimes something as innocuous as a change in indentation causes unexpected grief, which time it takes to deal with may be multiplied by the number of downstream users.  Even worse is when the tool in question tends to be used in CI pipelines where the latest version is used automatically, without the downstream users' having a chance to read changelogs and make changes for compatibility.

**Comment by meedstrom** [comment]
> And FWIW, I would not like this check to be disabled by default, because then I'd have to configure all of my packages to specifically enable it

You mean your initfiles, right? At least, I can't see putting a package-lint setting in dir-locals.el when one could just insert the Version: header and then that's it for that package, permanently.

As for makem.sh, I understand your concern about downstream users.  But can you describe a situation where it would cause grief? It is the removal of a message, not introduction of a CI-stopping step.  I'll buy that the removal of an *error* can feasibly break CIs, but this was never an error.

Suppose this check is removed today. Everyone who was warned up to today has already been warned and either heeded it or not; and new packages created since today do not get the warning, so at some point in the future, an individual tries `package-install-file`, gets a specific error from that command, fixes it by inserting a Version: header, and goes on their merry way.  Maybe they come here to file a suggestion, thinking it'd be a great idea for package-lint to warn about it, then @purcell says "nah", and they say "k". Case closed.

And yea I knew the Elisp manual says that, but it's written by the same people who do ELPA, so of course it would say that. It is also de facto incorrect: neither one nor the other "must" be present to reliably distribute a package, as we can see. 

I do not think they are positioned to know best on packaging matters.  I'd defer to MELPA maintainers as the most experienced authority, nevermind that they don't have commit rights to the Elisp manual and so are left without voice on a technicality. The Elisp manual could also tell you what party to vote for but you would be right to ignore that too.

**Comment by alphapapa** [comment]
@meedstrom Maybe I'm misunderstanding you, or maybe vice versa, I'm not sure.  Anyway, my point is that, if this check were disabled by default, it would no longer serve the purpose of ensuring that such a header is present; so in order to keep doing that check, downstream users would have to modify their tooling to enable it.

That would be especially inconvenient for a tool like `makem.sh` which may be used on new packages: its purpose is to find such problems ahead of time for the developer, not to defer finding them to installation time for the user.

I don't think your characterization of the Elisp manual or its maintainers is quite fair.  MELPA maintainers certainly have a voice, by using the mailing list or filing a bug report.  Some of the MELPA maintainers may actually be committers to emacs.git as well (AFAIK at least one is).

If there are deficiencies in the Elisp manual, I'd urge you to propose improvements to address them, rather than just criticizing them here.  You seem to have an impression of some kind of resentment between ELPA and MELPA, but I don't think that is the case.  The two repositories serve similar but different purposes.  Most contributors to ELPA are probably also contributors to MELPA, in some form.  We should not consider the two efforts to be opposed.

**Comment by meedstrom** [comment]
> not to defer finding them to installation time for the user.

This is well pointed out, if that's what would happen, but if I understand correctly, it will not.  Consider that package.el is probably used in the first place to install packages off either ELPA or MELPA. The first would catch the absence before hosting the package in the first place, and the second inserts the version header.  

Aftermarket package managers such as Elpaca do not execute `package-install-file` in the first place. (I'm not sure though how Quelpa/Borg/el-get work... we'd have to look into that if we want to carry out this suggestion, but we're not going to so it's all theoretical now.)

> I don't think your characterization of the Elisp manual or its maintainers is quite fair.

Okay, maybe I'm feeling hyperbolic because of the discussion on https://lists.gnu.org/archive/html/emacs-devel/2024-09/msg00366.html.

Also because that passage in the manual was the first thing I did see, so my packages used to have a Version:.  Me trying to do the right thing. Then the subject of that email threatened to cause breakage for my users - imagine how I felt when I had to rewrite the commit history to undo my mistake. Now I cannot trust what the manual says about The Right Thing (nor package-lint, for that matter). 

I very much agree with not deferring problems to users at installation time, which is *why* I think this warning should not be offered to developers.

Maybe you are right and the manual just has a bug that could be filed: its claims are too broad and could at least clarify the potential consequences, especially with Emacs 30.

My FSF assignment form is dragging, but I'll make a contribution in the future, emacs-devel willing.

>  it would no longer serve the purpose of ensuring that such a header is present; 

Indeed we might be talking past each other a little :) I tried to say that this check need not be a purpose of an Elisp package linter. I only offered "disable by default" instead of "delete" because you never know.

**Comment by meedstrom** [comment]
> AFAIU you only needed to push a new commit updating the version to something like 0.6.1 or 0.7.0.

Yes - and to keep doing that regularly forever.  Which I had already chosen against doing. I did not mind tagging versions, but I did not also want to remember to do that, it's a headache to keep them in sync.

> Alternatively, if it was really an issue with package-vc, then fixing package-vc (which is relatively new) would be the best course of action, rather than expecting maintainers and users of existing packages to adapt to it.

(I think the module in question is called vc-use-package, FWIW) 

I agree!  That is why I emailed them.  It looks like they will not consent to fixing it, beyond making it understand tags (which is great news and reduces the scope of the problem by maybe as much as half).

> AFAICT you're talking about a very, very specific issue in package-vc and a new, unreleased version of Emacs. I don't think that would justify universally disabling an option like this--so, yes, a bit hyperbolic.

I understand it seems niche at first glance. But although the devs may not have envisioned it, use-package :vc looks like it'll be a built-in contender for Straight, and people may end up using it in exactly the same way as they use Straight.

That is, they may not just use :vc every now and then for a special case package that cannot be found on MELPA, but for *every* package.  The way that works, then, is that ELPA/MELPA function only as recipe collections, not package hosts, and the user's machine does all the git clonings locally.

This then becomes very relevant, very quickly, as its defaults are applied to every package.

**Comment by alphapapa** [comment]
> > AFAIU you only needed to push a new commit updating the version to something like 0.6.1 or 0.7.0.
> 
> Yes - and to keep doing that regularly forever. Which I had already chosen against doing. I did not mind tagging versions, but I did not also want to remember to do that, it's a headache to keep them in sync.

As  has been pointed out, meaningfully versioning software is a best practice, and for anything non-trivial, it should be done.  But every maintainer may do as he pleases--and that includes Emacs's maintainers, who IMO are doing the right thing by not choosing to encourage poorer practices.

> 
> > Alternatively, if it was really an issue with package-vc, then fixing package-vc (which is relatively new) would be the best course of action, rather than expecting maintainers and users of existing packages to adapt to it.
> 
> (I think the module in question is called vc-use-package, FWIW)

I'm referring to the soon-to-be-built-in `package-vc` by Philip, not the `vc-use-package` third-party library.

> 
> I agree! That is why I emailed them. It looks like they will not consent to fixing it, beyond making it understand tags (which is great news and reduces the scope of the problem by maybe as much as half).
> 
> > AFAICT you're talking about a very, very specific issue in package-vc and a new, unreleased version of Emacs. I don't think that would justify universally disabling an option like this--so, yes, a bit hyperbolic.
> 
> I understand it seems niche at first glance. But although the devs may not have envisioned it, use-package :vc looks like it'll be a built-in contender for Straight, and people may end up using it in exactly the same way as they use Straight.
> 
> That is, they may not just use :vc every now and then for a special case package that cannot be found on MELPA, but for _every_ package. The way that works, then, is that ELPA/MELPA function only as recipe collections, not as download servers, and the user's machine does all the git clonings locally.
> 
> This then becomes very relevant, very quickly, as its defaults are applied to every package.

I'm afraid that that is just not Emacs's responsibility.  Straight is a third-party package with various idiosyncrasies (some of which have been confusing and problematic to users, from my perspective as a package maintainer, having received complaints involving its use).  It's good in many ways, but it's not necessarily a model for Emacs to follow.  If users choose to use it, and then choose not to, it should be expected that some adjustments may be needed.

**Comment by meedstrom** [comment]
I guess that our fundamental priorities differ then. From my perspective it seems as if you do not care about breaking things for users, but I'm sure you do! I just don't see how it works out. 

I've noticed the same thing with some other individuals, that it seems as though "The Right Thing" and "cause least breakage" is *not the same thing*. For me, it is. It does not matter what is the responsibility of Emacs or not. A move either causes breakage or not, and should be judged on that basis. Am I going mad?

**Comment by alphapapa** [comment]
> I guess that our fundamental priorities differ then. From my perspective it seems as if you do not care about breaking things for users, but I'm sure you do! I just don't see how it works out.

A strange perspective, indeed, considering how many hours I've spent avoiding such breakage in my projects, including developing tools like [makem.sh](https://github.com/alphapapa/makem.sh) to prevent such breakage from occurring.
> 
> I've noticed the same thing with some other individuals, that it seems as though "The Right Thing" and "cause least breakage" is _not the same thing_. For me, it is. It does not matter what is the responsibility of Emacs or not. A move either causes breakage or not, and should be judged on that basis. 

Responsibility is a fundamental principle.  No project can take responsibility for what other projects do; no person can take responsibility for what other people do; no developer can take responsibility for what downstream users do.  Imagine the madness that would ensue otherwise: there would be no boundary between what must be accounted for in a project and what is outside its purview.  Effectively, a software project would become one enormous project to account for every user and every need and everything in the universe.

A line must be drawn, and Emacs naturally draws it at third-party software and projects, including Straight, `vc-use-package`, and MELPA.

It's even more natural when you consider that the Elisp guidelines have said for years that a version header should be maintained.  MELPA chose to use that header differently, and various package authors choose to ignore the guidelines--and now you expect Emacs to conform to those who have ignored or rejected that advice?

> Am I going mad?

Well, you certainly seem to have unrealistic expectations for upstream projects.  :)

**Comment by meedstrom** [comment]
> No project can take responsibility for what other projects do; no person can take responsibility for what other people do; no developer can take responsibility for what downstream users do.

I do, and my projects do. It's easy: when it's brought to my attention, or I realize myself, that "X causes breakage when used with my package" I implement a workaround if it's easy, or else declare X as unsupported with some kind of helpful warning message.

Emacs 30 is currently set to do neither: break something that would be easy to keep unbroken, then don't warn those that would be most affected.  The current discussions about informing users with some kind of `sit-for` message at init would not help those who would be most affected.

**Comment by meedstrom** [comment]
> there would be no boundary between what must be accounted for in a project and what is outside its purview. 

There is no boundary :)  "Nihil supernum" - there's no intergalactic authority to judge you at the end of your life and say, yes, you did what was within your responsibility, so you can go to heaven.  You either knowingly break things or you don't, and then those things are either broken or not broken, as a result. If you don't want it broken, even though it'd be *totally someone else's fault*... the only thing that will get you what you want (it not being broken) is to not break it.  So it's all down to what outcome you wanted and how badly you wanted it. Does that make sense?

Responsibility doesn't really exist. There is only breakage.

**Comment by alphapapa** [comment]
> > No project can take responsibility for what other projects do; no person can take responsibility for what other people do; no developer can take responsibility for what downstream users do.
> 
> I do, and my projects do. It's easy: when it's brought to my attention - or I realize myself - that "X causes breakage when used with my package" I implement a workaround if it's easy, or else declare X as unsupported with some kind of helpful warning message.

So you'd potentially add infinite "X is unsupported" warning messages?  What is the limiting principle?  

> 
> Emacs 30 is currently set to do neither: break something that would be easy to keep unbroken, then don't warn those that would be most affected, even after having all this brought to their attention. The current discussions about informing users with some kind of `sit-for` message at init would not help those who would be most affected.

The idea of defining breakage as "anything a user might do that might be affected by a change we might make" does not scale, especially when we're talking about *new* functionality in a *new* version of the software.

Emacs does not take responsibility for ensuring compatibility with third-party add-ons.  It cannot.

> There is no boundary :) "Nihil supernum" - there's no intergalactic authority to judge you at the end of your life and say, yes, you did what was within your responsibility, so you can go to heaven. You either knowingly break things or you don't, and then those things are either broken or not broken, as a result. If you don't want it broken, even though it'd be totally someone else's fault... the only thing that will get you what you want (it not being broken) is to not break it. So it's all down to what outcome you wanted and how badly you wanted it. Does that make sense?

No, I don't think it does.  Your logic would effectively require mind-reading and future-telling.  It would make everyone responsible for everything everyone else does.  It doesn't jibe with reality.

> Responsibility doesn't really exist. There is only breakage.

That sounds nice, but even the Linux kernel, despite Linus's axiom that "WE DON'T BREAK USERSPACE" has limits: internal APIs are free to change, and out-of-tree modules are not the kernel's responsibility (i.e. they don't break userspace, but anything third-party that is not in userspace is not their responsibility).

Our most limiting factor is time; there are only 24 hours in a day, and only so many contributors to a project.  Projects define themselves and their responsibilities toward users; outside of them is the responsibility of others.  Short of sci-fi, Borg-like hive-mind, this is the reality of human existence.

I hesitate to say something like this, because I don't know much about you and your experience, but I do know that the Emacs maintainers have many years of experience maintaining very large software projects used by many, many users, and I have similar experience at a smaller scale--and I'd say that, if you were in their shoes for a little while, you'd quickly understand why they have their policies.

Ideals are inspirational, but only with pragmatism is progress made in the real world.

**Comment by meedstrom** [comment]
Beautiful words. Of course, I agree with \*checks notes\* all of that. Limited 24 hours, pragmatism necessary, and so on. But our takeaways still differ mightily, and it boils down to what responsibility means.  

It's a social construct for assigning blame.  

That's what we mean when we say we *take responsibility* - some sort of implicit message that *you can all frown at me if I fail at maintaining standards here*.

The concept also can be used to avoid blame.  Like a teacher that sees that a kid is being abused, but sees no obvious way to intervene, and stays out because it's not their responsibility anyway. Or they report it to some sort of "higher" authority like CPS, and then they have discharged their duty: it's in CPS' hands now.

In that sense... I take no responsibility for anything.  I fix bugs because it pleases me to do so, because it pleases me to publish non-broken code under my name, but that is the whole reason. I would disagree with someone who said I have a *responsibility* to fix my bugs. Still, I keep doing it. Because I want to.

But this can lead paradoxically, to doing *more* than less.  If I was a teacher in aforementioned situation, why, I might visit the kid's parents at their home if the situation offended me enough to compel me to do so. I would realize that CPS might not do an adequate job, and then the "responsibility" falls to me as far as I care about that kid - in the same way responsibility falls to me to prevent breakage for users, as far as I care about those users.

That is - I have no responsibility for these things whatsover. But the best way to bring about outcomes I care about is to *bring them about*. :)

In summary...  You are responsible for everything bad that happens in the world, Adam.

It's born with you and you cannot avoid it.

(It is also born with me... not that you can *blame* me for anything, nor I you. Blame is pointless.)

Even though you cannot read minds and predict the future, it is still your responsibility to read minds and predict the future, even create utopia. If you *want* utopia, no one else can be trusted with the responsibility.

But this is a wonderful thing! So many choices you can make, so many timelines that can play out differently as a direct consequence of your choices. All constrained by how smart you are, how far you can see, and how much you care. That's the great game of life, isn't it grand?


-----

All that to summarize that either emacs-devel got too wrapped up in *who has what responsibility* that they forgot what they care about, or they just don't particularly care about the users who would suffer due to this. And that's okay. I don't blame them, blame is pointless. I even think that if they find it unpleasant to contribute to Emacs, because of the *responsibilities*, it would make sense for them to quit entirely and spend their days golfing.  But I'm still here to try to convince who's left to flip this setting, because I care about those users more than I care about compelling every would-be Elisp packager to version their releases predictably. It really offends me to break people's initfiles, so this is *the* pragmatic setting, and it's even easy to do even though you only have 24 hours a day.

**Comment by meedstrom** [comment]
All that to summarize that either emacs-devel got too wrapped up in *who has what responsibility* that they forgot what they care about, or they just don't particularly care about the users who would suffer due to this. And that's okay. I don't blame them, blame is pointless. I even think that if they find it unpleasant to contribute to Emacs, because of the *responsibilities*, it would make sense for them to quit entirely and spend their days golfing.  But I'm still here to try to convince who's left to flip this setting, because I care about those users more than I care about compelling every would-be Elisp packager to version their releases predictably. It really offends me to break people's initfiles, so this is *the* pragmatic setting, and it's even easy to do even though you only have 24 hours a day.

**Comment by meedstrom** [comment]
You know, I was going to wrap all that up into some unassailable conclusion, but I started to feel that I was arguing myself into a corner that I don't actually agree with.  My words flowed easily and I couldn't find the fault in them, but it still felt strained, like there was a good chance I was going to have statements on my public record that sound moronic in hindsight.  

I prefer to short-circuit that process. 

So here's a couple cruxes that could totally change my views, if you'd like to engage. 

### 1. Safety

Do you agree that, in a situation where you cannot assume that the version header carries information (because people put it in regardless to pass a lint-check), it is safest to fetch master? I think you can see where I'm coming from, since you pointed out the issues with Melpa Stable a few years back. 

Fetching master is not a principle especially unique to Melpa, it is just the safe method and it's what I was saying makes sense for emacs core to do as well with :vc in order to not give :vc users a "Melpa Stable-style" instability, an instability which has the additional insult of potentially being difficult to trace back to the root cause and is therefore exceptionally unfriendly to users. 

Note, it is still possible to be a proponent of doing the unsafe thing even if one agrees it is unsafe (see point 3), but it it'd be interesting to know if you don't agree about what is safest currently.

### 2. Valid points that do not apply

Generally I agree with the principle that existing prior art should not be a reason to hold upstream back from developing as it wishes, but I think that this principle is being invoked in this discussion a lot as a knee-jerk reaction (not just by you), and does not seem to be applicable, because I can't see that there are any ambitions in upstream that would be blocked, if it opted to ship a reversed default. 
 
Does that make sense? It's like when you were talking about the merits of versioning in general, to which I agree, but it's moot because the topic is specifically about what to do with effectively unversioned packages that will stay that way for any reason imaginable or unimaginable.  If you can explain that actually upstream would be obstructed somehow, that would be helpful.

### 3. Treating any wild git repo as if it met ELPA standards

This is the biggest crux.

Why would it make sense for :vc to treat packages on a version-curated ELPA with an implicit versioning contract, the same way as packages not on such an ELPA?  

I.e. why would it make sense to presume that packages that have not opted in to any best practices *are* following best practices, when the consequence is that it harms not only the developers responsible but their users as well?

I guess that in the long term, a foolproof way to make an ecosystem of reliable versioned releases is to ensure that a developer must bump the version number in order to effectively distribute the new version to any users. 

I.e. so that it is not possible to ask one's package manager to download dev snapshots of *everything* by default, only specific ones by inserting URLs by hand or some such minor inconvenience. That's where emacs core can leverage its privileges of controlling the built-in defaults, it could really make this happen.

But I think it could still do that, in the future, even if it were to ship a safe default behavior today.

**Comment by alphapapa** [comment]
> That's what we mean when we say we _take responsibility_ - some sort of implicit message that _you can all frown at me if I fail at maintaining standards here_.

Yes, because, since we are not a Borg-like hive-mind, but rather independent actors, that is the only way that society can function: we specialize, doing what we are good at, for mutual benefit.  Otherwise we'd all be living in caves, each laboriously eking out our daily needs from scratch.

> In summary... You are responsible for everything bad that happens in the world, Adam.
> 
> It's born with you and you cannot avoid it.
> 
> (It is also born with me... not that you can _blame_ me for anything, nor I you. Blame is pointless.)
> 
> Even though you cannot read minds and predict the future, it is still your responsibility to read minds and predict the future, even create utopia. If you _want_ utopia, no one else can be trusted with the responsibility.

We clearly have very different philosophies (and that's wildly off-topic here; we should be considerate of our host, Steve).

> Do you agree that, in a situation where you cannot assume that the version header carries information (because people put it in regardless to pass a lint-check), it is safest to fetch master? I think you can see where I'm coming from, since you pointed out the issues with Melpa Stable a few years back.

I don't know what you mean by *safe* here.

The core problem is that each developer follows his own way: some are disciplined and methodical, some are free-wheeling.  Some packages are complex, have been around for years and are intended to be for many more, and change slowly; others are relatively trivial, may not require any changes for years, and might even be abandoned after a while, and then maybe picked up by someone else.  Some packages need careful versioning; others may need nothing more than git HEAD.

We cannot control what developers do; so a tool like `vc-package` can only be a tool to be applied with discretion by the user.  As Philip has said, he intended it as a tool for those who want to contribute to packages, not as a replacement for Straight, Elpaca, et al.  I think it could likely serve as both, but that would not obviate the need for users to apply it with discretion.  Users of Straight and Elpaca also should have some idea of what they are doing and how each package they install from git is developed; those who don't are frequently seen asking for help because they don't understand the consequences.  Not every package developer intends for his software to be installed directly from git by users who aren't themselves developers.  What we should really be doing is telling users that, if they choose to install a package directly from git, they should ensure that the developer recommends doing so, otherwise the user is on his own.

> Generally I agree with the principle that existing prior art should not be a reason to hold upstream back from developing as it wishes, but I think that this principle is being invoked in this discussion a lot as a knee-jerk reaction (not just by you), and does not seem to be applicable, because I can't see that there are any ambitions in upstream that would be blocked, if it opted to ship a reversed default.
> 
> Does that make sense? It's like when you were talking about the merits of versioning in general, to which I agree, but it's moot because the topic is specifically about what to do with effectively unversioned packages that will stay that way for any reason imaginable or unimaginable. If you can explain that actually upstream would be obstructed somehow, that would be helpful.

I don't think it's fair to describe those who disagree with you as having "knee-jerk reactions."  This position is principled and well-considered, based on decades of FOSS experience, individually and collectively.  The Emacs maintainers are careful not to obligate themselves and the project in unwise ways.  This is an example.  (If you categorically and cosmically reject the concept of responsibility, then there would seem to be no common ground on which to discuss it further.  It would then be for you to consider the consequences of applying such descriptors when debating these ideas.)

> Why would it make sense for :vc to treat packages on a version-curated ELPA with an implicit versioning contract, the same way as packages not on such an ELPA?

Emacs does not do such a thing.  `vc-package` is a tool; it is up to the users to apply it with wisdom.  It provides "mechanism, not policy."  If you want policy, use curated repositories like ELPA or MELPA.

**Comment by meedstrom** [comment]
You are being combative today, interpreting me in the worst possible lights, and I kind of understand why. I propose that we can this discussion rather than try to nurture it to health.

**Comment by alphapapa** [comment]
> You are being combative today, interpreting me in the worst possible lights

I don't think so.  You have been rude to me and to the Emacs maintainers, accusing us of thoughtlessly having unreasonable positions, among other things.  I have patiently explained why I disagree, and why these policies are in place.  I've been doing this to try to be helpful to you, since you don't seem to understand.

> I propose that we can this discussion rather than try to nurture it to health.

Sadly, that seems to be for the best.

**Comment by meedstrom** [comment]
Um, I'll have to review my emails. I definitely have to improve my communication. I thought I had been respectful and focused on just describing causes and effects as I saw them. As you demonstrate by perceiving me as rude/demanding, I failed to do that, and I am sorry.

It may also be that you and I were badly interacting chemical elements. I found the exchange mostly interesting up until a couple of posts ago, although I had experienced you as increasingly rude, or perhaps passionate. Now I am not sure if you had a good time at any point. Communication is so hard.

I do not request a continued discussion about the subject matter, but just for debugging, I will add that I did not experience that you put in any effort in your response https://github.com/purcell/package-lint/issues/275#issuecomment-2366822129, since you did not really answer 2 or 3, which I feel like you would have if you had tried to understand my point of view. I had been happy, thinking you would respond and I would finally understand and we could close the case, but then that response came. I felt attacked because you seemed to infer an accusing tone, which is not anywhere between my lines as I intended them.

**Comment by alphapapa** [comment]
Ok, for the sake of conversational debugging:

> Um, I'll have to review my emails. I definitely have to improve my communication. I thought I had been respectful and focused on just describing causes and effects as I saw them. As you demonstrate by perceiving me as rude/demanding, I failed to do that, and I am sorry.

> ...

> I felt attacked because you seemed to infer an accusing tone, which is not anywhere between my lines as I intended them.

It is kind of you to apologize.  FWIW, saying that those who disagree with you, who have carefully explained why they do, are having a "knee-jerk reaction"; and saying that it seems to you that I don't care about breaking things for users; things like that seem rude and accusatory to me.  I hope you see what I mean.

> It may also be that you and I were badly interacting chemical elements. I found the exchange mostly positive up until a couple of posts ago, although I had experienced you as increasingly rude, or perhaps passionate. Now I am not sure if you had a good time at any point. Communication is so hard.

I don't think you intended to be rude, but to me, you seemed somewhat rude from the beginning, including the wider context (i.e. emacs-devel, etc).  That's probably because you seemed to be demanding that the Emacs maintainers accede to your idea, despite their patient explanations.  And going into metaphysics doesn't help what should be primarily a technical discussion.

> I do not request a continued discussion about the subject matter, but just for debugging, I will add that I did not experience that you put in any effort in your response [#275 (comment)](https://github.com/purcell/package-lint/issues/275#issuecomment-2366822129), since you did not really answer 2 or 3, which I feel like you would have if you had tried to understand my point of view. I had been happy, thinking you would respond and I would finally understand and we could close the case, but then that response came. 

I did not write long paragraphs in response to your points 2 and 3, because my answers to them did not require it.  To say that I put no effort into them would be a case of mind-reading, which would be unfair; concise answers often require more effort than stream-of-consciousness writing.  I responded to what I interpreted as the points you were making.  Please consider reading them carefully alongside your questions, and you should see how the answers apply.

Again, my motivation here is to help you understand why your idea would not be appropriate to do in Emacs, and why Emacs's policies are what they are.  Since you seemed to push harder against those policies, I tried to explain further why those policies exist.  But given the apparent, fundamental difference in philosophy (re: responsibility, etc), I don't think there's much common ground on which to discuss the policies.  So I would just encourage you to consider how long Emacs has been around as a project, how successful it has been, and why it has.

**Comment by alphapapa** [comment]
@meedstrom As a final suggestion, I'd suggest that we both remove our comments here, since they are more like a personal conversation than for public consumption.  I'll remove mine now.

**Comment by purcell** [comment]
Catching up on the emails, I know what you guys spent your weekend on. 😂 Hope you both have a great week.

**Comment by purcell** [comment]
Don't worry, I'm gonna move on with my day and have another coffee instead.

**Comment by alphapapa** [comment]
> Don't worry, I'm gonna move on with my day and have another coffee instead.

Now that's good advice.  ;)  Thanks, Steve.
